### PR TITLE
BACK-32: Added the star history feature and transformed ScrapeRepos into a .ts file

### PR DESCRIPTION
**Pull request by 2015builder**
I've also updated the package.lock.json and the package.json for some imports.
The automatic assignement of API keys is not yet included in this update

### CODE REVIEW

File: ScrapeRepos.js [hunk 1]
```diff
@@ -0,0 +1,399 @@
+var __assign = (this && this.__assign) || function () {
```
Comment (pczern) [inline ScrapeRepos.js:1]: rewrite in TypeScript

**Review by pczern — CHANGES_REQUESTED** [review]
@2015builder ping me when converted file

File: package.json [hunk 2]
```diff
@@ -25,15 +25,21 @@
   },
   "homepage": "https://github.com/jst-seminar-rostlab-tum/truffle-ai-backend#readme",
   "dependencies": {
+    "@types/axios": "^0.14.0",
```
Comment (DogukanGun) [inline package.json:28]: Why there is two different implementations like @types/axios and axios

File: ScrapeRepos.ts [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+const cheerio = require('cheerio');
+const time = require('console');
+const showdown = require('showdown');
+
+
+/** Fetches the repo owner and name for each repo on the GitHub trending page
+ * @param {number} timeMode 0: daily; 1: weekly; 2: monthly
+ * @returns {string[]} an array that stores alternatingly the owner and the name of each repo
+ */
+async function fetchRepos( timeMode: number) {
```
Comment (DogukanGun) [inline ScrapeRepos.ts:11]: Please beatify your code

File: ScrapeRepos.ts [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+  try {
+    const timeModes = ['daily', 'weekly', 'monthly'];
+    const response = await axios.get('https://github.com/trending?since=' + timeModes[timeMode]);
+    const $ = cheerio.load(response.data);
+    const repos: string[] = [];
+
+    $('h2 a').each((i: number, el: cheerio.Element) => {
```
Comment (DogukanGun) [inline ScrapeRepos.ts:18]: h2 is to static. There would be new h2 tags if the design of page changed. Please come up with a new idea. May be id of html tags can be used or an api can be used.

File: ScrapeRepos.ts [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+      const repoName = $(el).text().trim();
+      repos.push(repoName);
+    });
+    
+    var trendingSplit: string[] = [];
+    // trim the repos to be correctly formatted
+    repos.forEach((repo) => {
+      let trimmedName = repo.replace(/\n\s+/g, '');
```
Comment (DogukanGun) [inline ScrapeRepos.ts:26]: You can use two replace in a row therefore you don't have to write this code into two lines.

File: ScrapeRepos.ts [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+      trimmedName = trimmedName.replace(/\//g, "");
+      const stringSplit = trimmedName.split(" ");
+      trendingSplit.push(stringSplit[0]);
+      trendingSplit.push(stringSplit[1]);
+    });
+
```
Comment (DogukanGun) [inline ScrapeRepos.ts:32]: Extra space is not needed.

File: ScrapeRepos.ts [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+    return trendingSplit;
+  } catch (error) {
+    console.error(error);
```
Comment (DogukanGun) [inline ScrapeRepos.ts:35]: For the security purpose, logs should not be printed.

File: ScrapeRepos.ts [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+      // Use the converter object to convert Markdown to HTML to String:
+      const html = converter.makeHtml(content);
+      const html2 = html.toString();
+      const text = html2.replace(/<[^>]*>/g, '');
+      const text2 = text.replace(/\n\s+/g, '');
+      const text3 = text2.replace(/\//g, "");
+      return text3;
+    }
+    catch(error){
+
```
Comment (DogukanGun) [inline ScrapeRepos.ts:68]: Extra space is not needed

File: ScrapeRepos.ts [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+      const text = html2.replace(/<[^>]*>/g, '');
+      const text2 = text.replace(/\n\s+/g, '');
+      const text3 = text2.replace(/\//g, "");
+      return text3;
+    }
+    catch(error){
+      
+    }
+  }
+  console.log("ReadMe couldn't be found")
```
Comment (DogukanGun) [inline ScrapeRepos.ts:71]: Please throw an exception instead of printing.

File: ScrapeRepos.ts [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+async function getRepoInfo(query: String, authToken: String){
+  try{
+    const response = await  axios.post('https://api.github.com/graphql', {
+      query: query
+    }, {
+      headers: {
+        'Authorization': authToken 
+      }
+    });
+    const output = response.data.data.repository;
```
Comment (DogukanGun) [inline ScrapeRepos.ts:90]: Please return response.data.data.repository directly. Don't need to create a variable.

File: ScrapeRepos.ts [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+}
+
+
+/**  This function imports the ReadMe.md file for a repository (if it can be located)
+* @param {string} owner - owner of the repo
+* @param {string} name - name of the repo
+* @returns {string} a string containing the text of the repo. Returns null if the repo can't be located
+*/
+async function getReadme(owner: String, name: String) {
+  // these paths exists to check in multiple locations for the readme files
```
Comment (DogukanGun) [inline ScrapeRepos.ts:46]: Please add description on top of the function not inside the function if the description is not about something in the function

File: ScrapeRepos.ts [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+        },
+        {
+          role: "user",
+          content: question,
+        },
+      ],
+    }),
+  });
+  const data = await response.json();
+  const content = data.choices[0].message.content;
```
Comment (DogukanGun) [inline ScrapeRepos.ts:132]: Content variable is not needed to define. Please directly return the result.

File: ScrapeRepos.ts [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+  `;
+
+  const headers = {
+    'Authorization': authToken
+  };
+
+  let hasNextPage = true;
+  let endCursor: string | null = null;
+
+  let starHistoryList: any[] = []
```
Comment (DogukanGun) [inline ScrapeRepos.ts:172]: Please do not use any. Never can be option.

File: ScrapeRepos.ts [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+      starHistoryList.push(starsData)
+
+      const pageInfo = response.data.data.repository.stargazers.pageInfo;
+      hasNextPage = pageInfo.hasNextPage;
+      endCursor = pageInfo.endCursor || null;
+
+      variables.cursor = endCursor;
+    }
+  } catch (error) {
+    console.error('Error:', error);
```
Comment (DogukanGun) [inline ScrapeRepos.ts:197]: Please don't print the error.

File: ScrapeRepos.ts [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+    return []
+  }
+
+  // get the real star history
+  const mergedArray = [].concat(...starHistoryList);
+  const result = await starHistory(mergedArray);
```
Comment (DogukanGun) [inline ScrapeRepos.ts:203]: Please directly return the result of function. You should not create a new variable.

File: ScrapeRepos.ts [hunk 16]
```diff
@@ -0,0 +1,356 @@
...
+
+
+/** Get trending developers (and their trending repos) from the github page
+ * @param {number} userInput describes the timeframe; 0: daily, 1: weekly; 2: monthly
+ * @returns   list of {name: 'NAME', username: 'USERNAME', repo: 'REPO'}
+ */
+async function fetchDevelopers(userInput: number){
+  const timeModes = ['daily', 'weekly', 'monthly'];
+
+  interface Developer {
```
Comment (DogukanGun) [inline ScrapeRepos.ts:236]: Please don't define an interface inside a function. If possible define the interfaces into a new file.

File: ScrapeRepos.ts [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+  }
+
+  interface DeveloperRepo {
+    username: string |undefined;
+    repo: string;
+  }
+
+  axios.get('https://github.com/trending/developers?since=' + timeModes[userInput]).then((response: { data: any; }) => {
+    const html = response.data;
+    const $ = cheerio.load(html);
```
Comment (DogukanGun) [inline ScrapeRepos.ts:248]: Could you write a comment why you use dollar sign as a variable name.

File: ScrapeRepos.ts [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+    const developers: Developer[] = [];
+    const developerRepos: DeveloperRepo[] = []
+
+    // get the developer names and usernames
+    $('h1.h3.lh-condensed a').each((i: number, el: cheerio.Element) => {
```
Comment (DogukanGun) [inline ScrapeRepos.ts:253]: Please don't use a static html tag, you should use a class which is not used generally like bootstrap classes, or id

File: ScrapeRepos.ts [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+      const name: string = $(el).text().trim();
+      const username: string | undefined = $(el).attr('href')?.substr(1);
```
Comment (DogukanGun) [inline ScrapeRepos.ts:255]: Please do not use undefined.

File: ScrapeRepos.ts [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+    // correctly merge the two arrays
+    const repo = "";
+    const developersWithRepos = developers.map(developer => {
+      const matchingRepo = developerRepos.find(repo => repo.username === developer.username);
+      return { ...developer, ...(matchingRepo || {repo}) };
+    });
+
+    return developersWithRepos
+  }).catch((error: any) => {
+    console.log(error);
```
Comment (DogukanGun) [inline ScrapeRepos.ts:280]: .

File: ScrapeRepos.ts [hunk 21]
```diff
@@ -0,0 +1,356 @@
...
+
+    // get the developer names and usernames
+    $('h1.h3.lh-condensed a').each((i: number, el: cheerio.Element) => {
+      const name: string = $(el).text().trim();
+      const username: string | undefined = $(el).attr('href')?.substr(1);
+      developers.push({ name, username });
+    });
+
+    // get the repo name
+    $('h1.h4.lh-condensed a').each((i: number, el: cheerio.Element) => {
```
Comment (DogukanGun) [inline ScrapeRepos.ts:260]: .

File: ScrapeRepos.ts [hunk 22]
```diff
@@ -0,0 +1,356 @@
...
+
+
+/** Main function to test the functionality of the different methods
+ * and how to correctly call them and what the intended workflow is about
+ * @param Github and OpenAI token still have to be added manually
+ */ 
+async function main() {
+  // Choose whether to scrape daily, weekly or monthly information
+  const timeMode = 0
+  console.log("TimeMode: " + timeMode)
```
Comment (DogukanGun) [inline ScrapeRepos.ts:292]: .

File: ScrapeRepos.ts [hunk 23]
```diff
@@ -0,0 +1,356 @@
...
+
+  const trendingSplit: string[] | undefined= await fetchRepos(timeMode);
```
Comment (DogukanGun) [inline ScrapeRepos.ts:294]: .

File: ScrapeRepos.ts [hunk 24]
```diff
@@ -0,0 +1,356 @@
...
+  
+  // your personal GitHub authToken
+  const authToken = `Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX`;
+  // //Authorization: "Bearer " + process.env.OPENAI_API_KEY,
+  const OPENAI_API_KEY = "sk-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX";
+  
+  // check if any repos were actually found
+  if(!trendingSplit){
+      console.log("Repos couldn't be scraped")
```
Comment (DogukanGun) [inline ScrapeRepos.ts:303]: .

File: ScrapeRepos.ts [hunk 25]
```diff
@@ -0,0 +1,356 @@
...
+          totalCount
+        }
+        primaryLanguage {
+          name
+        }
+      }
+    }`;
+
+    const output = await getRepoInfo(query, authToken)
+    console.log(output)
```
Comment (DogukanGun) [inline ScrapeRepos.ts:329]: .

File: ScrapeRepos.ts [hunk 26]
```diff
@@ -0,0 +1,356 @@
...
+
+    // check if the repo has more than a 1k stars
+    const stars = output.stargazers.totalCount
+    if(stars < 1000){
+      console.log("Wouldn't provide this entity to the frontend")
```
Comment (DogukanGun) [inline ScrapeRepos.ts:334]: .

File: ScrapeRepos.ts [hunk 27]
```diff
@@ -0,0 +1,356 @@
...
+    if(readme != null){
+      // call openai api
+      const answer = await openAIRequestTurbo(readme, OPENAI_API_KEY)
+      console.log(answer)
+    }
+
+    // get the star history of the repo
+    // await fetchRepositoryStars(owner, name, authToken)
+    const result = await fetchRepositoryStars(owner, name, authToken)
+    console.log(result)
```
Comment (DogukanGun) [inline ScrapeRepos.ts:348]: .

File: ScrapeRepos.ts [hunk 28]
```diff
@@ -0,0 +1,356 @@
...
+  }
+  // get the developers
+  const developers = await fetchDevelopers(timeMode)
+  console.log(developers)
+  
+}
+main()
```
Comment (DogukanGun) [inline ScrapeRepos.ts:355]: Why did you use main? If you put this here for testing, please remove before opening the pr.

File: ScrapeRepos.ts [hunk 29]
```diff
@@ -0,0 +1,356 @@
...
+
```
Comment (DogukanGun) [inline ScrapeRepos.ts:356]: You should write your unit tests to test those functions.

File: ScrapeRepos.js [hunk 30]
```diff
@@ -0,0 +1,399 @@
...
+    __assign = Object.assign || function(t) {
+        for (var s, i = 1, n = arguments.length; i < n; i++) {
+            s = arguments[i];
+            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
```
Comment (DogukanGun) [inline ScrapeRepos.js:5]: If ts file is gonna be used, you should delete js file from pr.

**Review by DogukanGun — CHANGES_REQUESTED** [review]
Please add your unit tests.

File: ScrapeRepos.ts (hunk 3 above)
Comment (2015builder) [reply ScrapeRepos.ts:11]: Do you mean just rewriting the description as I did here or what do you mean?

File: ScrapeRepos.ts (hunk 4 above)
Comment (2015builder) [reply ScrapeRepos.ts:18]: The issue is that Github's API doesn't provide any information about its trending page. 
Therefore all the data wanted has to be received through scraping the HTML version of the file.

File: ScrapeRepos.ts (hunk 7 above)
Comment (2015builder) [reply ScrapeRepos.ts:35]: Should we already implement this? It will be harder to actually find errors without printing them where they origniated from. 
But if you think it's a good idea I also understand why and I will implement it

Comment (2015builder) [reply ScrapeRepos.ts:35]: Should we already implement this? The errors could be much harder to understand when not printed where they originated from.
But if you are sure that this is a good idea I understand why and I can implement this throughout the code.

File: ScrapeRepos.ts (hunk 29 above)
Comment (2015builder) [reply ScrapeRepos.ts:356]: So is this something you want me to do? And if so could you specifiy if I should use a framework for that; create a new file for that; and test every method deeply or just the main functionality?

File: ScrapeRepos.ts (hunk 4 above)
Comment (2015builder) [reply ScrapeRepos.ts:18]: I'be looked at the HTML code of the trending page for a bit but I am unsure which id/classname/tag would be the correct to choose from. Could you maybe have a look at it and tell me on which one to build from so that I can implement it here?

File: ScrapeRepos.js (hunk 1 above)
Comment (2015builder) [reply ScrapeRepos.js]: I just deleted it and kept the typescript version of the file

File: package.json (hunk 2 above)
Comment (2015builder) [reply package.json:28]: I didn't mean to add them. Sorry

**Comment by pczern** [comment]
Will this PR be affected by rate limiting? Could we ban our server from GitHub?

**Comment by 2015builder** [comment]
> Will this PR be affected by rate limiting? Could we ban ourselves from fetching repos? It might be better to use the API.

@pczern I can't say for sure whether we are affected by this but during my own tests I've never experienced any issues with that. FYI the Github API doesn't provide any information about the Github trending page. Therefore there is not really another way than scraping the Github trending page through looking at the HTML and then get the respective information for each repo through the Github API because this can be requested

**Comment by pczern** [comment]
Indeed there is @2015builder 
https://stackoverflow.com/questions/30525330/how-to-get-list-of-trending-github-repositories-by-github-api says that GitHub just uses their API
On this page, you get an introduction how to fetch data with GraphQL API https://katopz.medium.com/how-to-search-with-github-graphql-e6c142dc61ed
The rate limit will be much higher, if we follow this approach and it will be much cleaner

File: ScrapeRepos.ts (hunk 3 above)
Comment (DogukanGun) [reply ScrapeRepos.ts]: I mean reformat the code. For example there is an extra space between parantheses and timeMode. It should be like fetchRepos(timeMode: number)

**Comment by 2015builder** [comment]
> Indeed there is @2015builder https://stackoverflow.com/questions/30525330/how-to-get-list-of-trending-github-repositories-by-github-api says that GitHub just uses their API On this page, you get an introduction how to fetch data with GraphQL API https://katopz.medium.com/how-to-search-with-github-graphql-e6c142dc61ed The rate limit will be much higher, if we follow this approach and it will be much cleaner

@pczern I've already looked at the same StackOverflow page that you just sent me. It clearly states there that the GitHub API doesn't provide any information about the content of the GitHub trending page. So the only way to retrieve this data that I've found and know about is to scrape the HTML of the webpage and then to extract the Repos and Owners through search in this HTML code.
Of course one could also use the technique described in the medium.com article that you have sent me where one can retrieve simply all repositories above a specific threshold (e.g. 10k stars) but this wouldn't be helpful to us because this isnt focused on new, upcoming startups that could be early stage startups. Scraping all Github repositories with more than 10k stars would provide us with mostly very big (contributor wise) and older (therefore more stars) repositories that wouldn't necessarily fall into the category that La Famiglia is looking for.
So in summary I still believe that there is no good solution other than the one that I've currently implemented but I am totally open if you find a better method.

File: ScrapeRepos.ts (hunk 4 above)
Comment (2015builder) [reply ScrapeRepos.ts]: @DogukanGun would something like '$$('ol.repo-list li h3').forEach(el => console.log(el.innerText));' this work better as a way too extract repos and owners. Because that's something I could definitely implement

**Comment by pczern** [comment]
@2015builder you are right
I didn't see it initially, it seemed to me to be the same.
I am sorry, your approach is really good

**Comment by 2015builder** [comment]
I have now overwritten the star history of my code and because this was such a huge change I talked with Philipp about it and I think it is best to close this PR and to open a new one on the same branch
